### PR TITLE
Update callbacks add audio feedback

### PR DESCRIPTION
**Pull request by lduffy-voysis**
- replaced call(event: Event) with individual callbacks. success/failure are mandatory, the rest are optional.
- Added FinishedReason enum used by recordingFinished() method.
- Added audioData callback containing recorded data.
- removed generics from serviceImpl, converted startAudioQuery to generic method.
- removed global callback from serviceImpl
- refactored serviceImpl. Removed labels from method.

### CODE REVIEW

File: Voysis/events/Callback.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import Foundation
+
+public protocol Callback: class {
```
Comment (jfarrelly-voysis) [inline Voysis/events/Callback.swift]: Apparently `: AnyObject` is the preferred way to do class-only protocols since Swift 4.

Comment (lduffy-voysis) [reply Voysis/events/Callback.swift]: sounds good. Updated.

File: Voysis/events/CallbackDispatcher.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import Foundation
+
+internal class CallbackDispatcher<T: Callback> {
+    private let dispatchQueue: DispatchQueue
+    private weak var callback: T?
```
Comment (jfarrelly-voysis) [inline Voysis/events/CallbackDispatcher.swift]: This should be mandatory?

File: Voysis/events/CallbackDispatcher.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import Foundation
+
+internal class CallbackDispatcher<T: Callback> {
+    private let dispatchQueue: DispatchQueue
+    private weak var callback: T!
```
[the remark this answers is not in the document]
Comment (lduffy-voysis) [reply Voysis/events/CallbackDispatcher.swift]: oh yeah. good point. removed.

**Pull request merged**